layer1/executor/task_executor.go

```diff
@@ -0,0 +1,354 @@
+				continue
+			}
+			logger.Debugf("got a unrecoverable error during task.execute finishing execution err: %v", taskErr.Error())
+			return taskErr
+		}
+		if txn != nil {
+			logger.Debugf("got a successful txn: %v", txn.Hash().Hex())
+			err := te.addTxBackup(task.GetId(), txn)
+			if err != nil {
+				return err
+			}
+
+			isComplete, err := te.checkCompletion(ctx, task, txn)
+			if err != nil {
+				return err
+			}
+			if isComplete {
+				return nil
+			}
+			continue
+		}
+		logger.Debug("Task returned no transaction, finishing")
+		return nil
+	}
+}
+
+// checkCompletion checks if a task is complete. The function is going to subscribe a
+// transaction in the txWatcher, and it will wait until it gets the receipt,
+// the task is killed, or shouldExecute returns false.
+func (te *TaskExecutor) checkCompletion(ctx context.Context, task tasks.Task, txn *types.Transaction) (bool, error) {
+	var err error
+	var receipt *types.Receipt
+	logger := task.GetLogger()
+
+	receiptResponse, err := te.txWatcher.Subscribe(ctx, txn, task.GetSubscribeOptions())
+	if err != nil {
+		// if we get an error here, it means that we have a corrupted txn we should
+		// retry a transaction
+		logger.Errorf("failed to subscribe tx with error: %s", err.Error())
+		return false, nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return true, ctx.Err()
+		case <-time.After(constants.TaskExecutorPoolingTime):
+		}
+
+		if receiptResponse.IsReady() {
+			receipt, err = receiptResponse.GetReceiptBlocking(ctx)
+			if err != nil {
+				var txnStaleError *transaction.ErrTransactionStale
+				if errors.As(err, &txnStaleError) {
+					logger.Info("got a stale transaction and couldn't retry, finishing execution")
+					return false, err
+				}
+				logger.Warnf("got a error while waiting for receipt %v, retrying execution", err)
+				return false, nil
+			}
+
+			if receipt.Status == types.ReceiptStatusSuccessful {
+				logger.Debug("got a successful receipt")
+				return true, nil
+			} else {
+				logger.Warn("got a reverted receipt, retrying")
+				return false, nil
+			}
+		}
+
+		logger.Trace("receipt is not ready yet")
+		hasToExecute, err := shouldExecute(ctx, task)
+		if err != nil {
+			return false, err
+		}
+		if !hasToExecute {
+			return true, nil
+		}
+	}
+}
+
+// shouldExecute checks if a task should be executed. In case of recoverable errors the
+// function is going to retry `constants.MonitorRetryCount` times. If it
+// exhausted the number of retries, it sends true since there's no information.
+// The function returns false in case of unrecoverable errors, or if it
+// shouldn't execute a task. In case of no errors, the return value is true
+// (default case to t.ShouldExecute to return that a task should be executed).
+func shouldExecute(ctx context.Context, task tasks.Task) (bool, error) {
+	logger := task.GetLogger()
+	for i := uint64(1); i <= constants.MonitorRetryCount; i++ {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
```

Comment:
Cover in tests if possible.

Reply:
Couldn't cover. If I kill the context during the test, the execution finishes before this line with another context check.